cpp/lib/include/Xerces.h

```diff
@@ -0,0 +1,123 @@
+/** \file   Xerces.h
+ *  \brief  Wrapper class for Xerces XML parser
+ *  \author Mario Trojan (mario.trojan@uni-tuebingen.de)
+ *
+ *  \copyright 2018 Universitätsbibliothek Tübingen.  All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero %General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <memory>
+#include <queue>
+#include <string>
+#include <vector>
+#include <xercesc/framework/XMLPScanToken.hpp>
+#include <xercesc/parsers/SAXParser.hpp>
+#include <xercesc/sax/HandlerBase.hpp>
+#include <xercesc/util/XMLString.hpp>
+#include "util.h"
+
+class Xerces {
+    xercesc::SAXParser * parser_;
+    xercesc::XMLPScanToken token_;
+    std::string xml_file_;
+    bool prolog_parsing_done_ = false;
+    bool body_has_more_contents_;
+public:
+    struct XmlPart {
+        enum Type { START_ELEMENT, END_ELEMENT, PROCESSING_INSTRUCTION, CHARACTERS, IGNORABLE_WHITESPACE };
+        virtual Type getType() const = 0;
+        virtual ~XmlPart() = default;
+        static std::string TypeToString(const Type type);
+    };
+
+    struct StartElement : XmlPart {
+        std::string name_;
+        std::vector<std::pair<std::string, std::string>> attributes_;
+        inline virtual Type getType() const override { return START_ELEMENT; }
+    };
+
+    struct EndElement : XmlPart {
+        std::string name_;
+        inline virtual Type getType() const override { return END_ELEMENT; }
+    };
+
+    struct ProcessingInstruction : XmlPart {
```

Comment:
Processing instructions must be handled internally and not exposed to this class' clients!